lib/api/model/model.dart

```diff
@@ -305,6 +307,121 @@ sealed class Message {
   Map<String, dynamic> toJson();
 }
 
+/// A message's reactions, in a convenient data structure.
+class Reactions {
+  final List<Reaction> unaggregated;
+
+  /// A list of [ReactionWithVotes] objects.
+  ///
+  /// There won't be two items with the same
+  /// [ReactionWithVotes.reactionType] and [ReactionWithVotes.emojiCode].
+  /// (We don't also key on [ReactionWithVotes.emojiName];
+  /// see [ReactionWithVotes].)
+  late final List<ReactionWithVotes> aggregated;
+
+  Reactions(this.unaggregated) {
+    final byReaction = <int, ReactionWithVotes>{};
+    for (final reaction in unaggregated) {
+      final key = Object.hash(reaction.reactionType, reaction.emojiCode);
+      final current = byReaction[key] ??= ReactionWithVotes(
+        reactionType: reaction.reactionType,
+        emojiCode: reaction.emojiCode,
+        emojiName: reaction.emojiName,
+      );
+      current.addVote(reaction.userId);
+    }
+
+    aggregated = byReaction.values.sorted(
+      // Descending by number of votes
+      (a, b) => -a._userIds.length.compareTo(b._userIds.length),
+    );
+  }
+
+  void add(Reaction reaction) {
+    unaggregated.add(reaction);
+    final current = aggregated.where((r) {
```

Comment:
This linear-scan approach seems fine. In principle it could be slow, if the message has a long list of distinct reactions; but I think that's very uncommon.

In the constructor, though, good to use a hash table, as you do. That's potentially processing quite a lot of reactions at once (even though usually not that many distinct ones), plus it can be part of processing a whole bunch of messages at once. So the same speed difference there would have a much greater risk of causing a dropped frame.